lib/spack/spack/user_environment.py

```diff
@@ -62,40 +68,57 @@ def unconditional_environment_modifications(view):
     return env
 
 
-def environment_modifications_for_spec(spec, view=None, set_package_py_globals=True):
+@contextmanager
+def projected_prefix(specs, projection):
+    prefixes = dict()
+    for s in traverse.traverse_nodes(specs, key=lambda s: s.dag_hash()):
+        if s.external:
+            continue
+        prefixes[s.dag_hash()] = s.prefix
+        s.prefix = prefix.Prefix(projection(s))
+
+    yield
+
+    for s in traverse.traverse_nodes(specs, key=lambda s: s.dag_hash()):
+        s.prefix = prefixes.get(s.dag_hash(), s.prefix)
+
+
+def environment_modifications_for_specs(
+    specs: Union[spack.spec.Spec, List[spack.spec.Spec]], view=None, set_package_py_globals=True
+):
```

Comment:
Can we use the following API:
```suggestion
def environment_modifications_for_specs(
    *specs: spack.spec.Spec, view=None, set_package_py_globals=True
):
```
? That will force setting the `view` and `setup_package_py_globals` argument by keyword (which is good imo) and make it unnecessary to have the `isinstance`check at the beginning of the function.

Reply:
It's also a footgun, cause `f(a, b)` has to be replaced with `f(a, b=b)`, but yeah, I went for this.